### PR TITLE
Implement feature #8: Add expense input validation and success message

### PR DESCRIPTION
**Pull request by Kimhyewon0621**
This PR solves #8 

- Integrated readRequiredLine and readBigDecimal methods into the addExpenseFlow to enhance input validation for required fields and the amount (rejects empty strings, non-numeric values, zero, and negative amounts).

- Added a success message (SUCCESS: Expense added.) to the end of the addExpenseFlow upon successful validation and (intended) storage.

- Added unit tests (AppTest.java) 

The call to store.append(DATA_FILE, exp); within addExpenseFlow in App.java has been temporarily commented out. This is because the ExpenseStore.append method is not yet implemented and was causing the tests to fail.

### CODE REVIEW

**Comment by hayat711** [comment]
**Group4** will be reviewing this PR [code implemented by group5 for feature #8 ]

File: src/main/java/itm/oss/splitter/App.java [hunk 1]
```diff
@@ -66,20 +66,25 @@ private void ensureDataDir() {
...
+    //String category = readLine("Category: ");
+    String category = readRequiredLine("Category: ");
     String notes = readLine("Notes (optional): ");
 
     Expense exp = new Expense(date, payer, amount, currency, participants, category, notes);
     ExpenseValidator.validate(exp);
-    store.append(DATA_FILE, exp);
-    System.out.println("Expense added.");
+    //store.append(DATA_FILE, exp);
+    System.out.println("SUCCESS: Expense added.");
```
Comment (ERE252) [inline src/main/java/itm/oss/splitter/App.java:88]: Nice job reusing readRequiredLine and readBigDecimal for all the required fields in addExpenseFlow. This makes the input validation much clearer and keeps the logic in one place instead of duplicating it inline. As a small suggestion, it might be helpful to mention in the success message that the input passed both field validation and ExpenseValidator.validate(exp), so users understand what “SUCCESS” means.

File: src/main/java/itm/oss/splitter/App.java [hunk 2]
```diff
@@ -208,4 +225,24 @@ private String join(ArrayList<String> xs, String sep) {
       r = r + sep + xs.get(i);
     return r;
   }
+  
+  BigDecimal readBigDecimal (String prompt) {
```
Comment (ERE252) [inline src/main/java/itm/oss/splitter/App.java]: The `readBigDecimal` method works well for validating numeric input and rejecting zero or negative values. 
Two minor suggestions:

(1) For consistency with other methods, consider removing the extra space in the signature (`BigDecimal readBigDecimal(String prompt)`).

(2) It could also help to add a short comment or Javadoc explaining that this method enforces “amount > 0”, since the method name alone doesn’t make that rule obvious.

File: src/test/java/itm/oss/splitter/AppTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    private App createAppWithSimulatedInput(String input) {
+        System.setIn(new ByteArrayInputStream(input.getBytes()));
+        return new App();
+    }
+
+    @Test
+    @DisplayName("Edge Case: readRequiredLine rejects empty input")
+    void testReadRequiredLine_EmptyInput_ShowsError(){
+        ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+        System.setOut(new PrintStream(outContent));System.setOut(new PrintStream(outContent));
```
Comment (yeonlimee2) [inline src/test/java/itm/oss/splitter/AppTest.java]: There are two consecutive calls to `System.setOut(new PrintStream(outContent));` on the same line.  
The second one immediately overrides the first, so it’s redundant.  

So I recommend remove redundancy for better readability.

```suggestion
        System.setOut(new PrintStream(outContent));

File: src/test/java/itm/oss/splitter/AppTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import org.junit.jupiter.api.DisplayName;
+import java.io.ByteArrayInputStream;
+
+public class AppTest {
+
+    // JUnit 5 기본 구조
+    @Test
+    void testApp_Placeholder() {
+        assertTrue(true); 
+    }
```
Comment (yeonlimee2) [inline src/test/java/itm/oss/splitter/AppTest.java]: The current test method testApp_Placeholder() does not show what it is testing.
It only works as a simple placeholder and does not explain its purpose.

Giving it a clearer name will make the test easier to understand.

```suggestion
    @Test  
    void testAppInitialization() {  
        assertTrue(true);  
    }

**Review by yeonlimee2 — COMMENTED** [review]
I left two small suggestions for clarity and readability:  
    1. Remove duplicate System.setOut() call in AppTest.java  
    2. Rename the placeholder test to show its purpose  
Both are minor, but they make the code cleaner and easier to understand.

**Comment by jobregon1212** [comment]
Hello guys, this PR should only modify the .java files (implementation and test files). It should not touch .gitignore , README, or any other file. So please fix that. Remember to synchronize with upstream repository and pull the latest changes from the upstream repository.

File: src/test/java/itm/oss/splitter/AppTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            assertFalse(output.contains("Error:")); 
+            assertTrue(output.contains("SUCCESS: Expense added.")); 
+
+        } finally {
+            System.setIn(originalIn);
+            System.setOut(originalOut);
+        }
+    }
+
+}
```
Comment (Jeffrey-Hong1005) [inline src/test/java/itm/oss/splitter/AppTest.java]: Nice work covering multiple input cases. 
Since System.setOut and ByteArrayOutputStream setup/restore blockes are repeated in every test, consider extracting them into @BeforeEach and @AfterEach methods to reduce duplication and make the tests easier to maintain.

```suggestion 

private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
private PrintStream originalOut;
private InputStream originalIn;

@BeforeEach
void setUpStreams() {
    originalOut = System.out;
    originalIn = System.in;
    System.setOut(new PrintStream(outContent));
}

@AfterEach
void restoreStreams() {
    System.setOut(originalOut);
    System.setIn(originalIn);
    outContent.reset(); // 다음 테스트 간 출력 겹침 방지
}

@Test
@DisplayName("Edge Case: readBigDecimal rejects invalid amounts")
void testReadBigDecimal_InvalidAmount_ShowsError() {
    App app = createAppWithSimulatedInput("abc\n0\n-100\n50.5\n");

    BigDecimal result = app.readBigDecimal("Enter amount: ");
    String output = outContent.toString();

    assertEquals(new BigDecimal("50.5"), result);
    assertTrue(output.contains("Please enter a valid number."));
    assertTrue(output.contains("Amount must be greater than zero"));
}
```

File: src/test/java/itm/oss/splitter/AppTest.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+        System.setOut(new PrintStream(outContent));
+
+        try{
+            App app = createAppWithSimulatedInput("abc\n0\n-100\n50.5\n");
+
+            BigDecimal result = app.readBigDecimal("Enter amount: ");
+            String output = outContent.toString();
+
+            assertEquals(new BigDecimal("50.5"), result);
```
Comment (Jeffrey-Hong1005) [inline src/test/java/itm/oss/splitter/AppTest.java]: Good job verifying that the valid numeric input is correctly parsed. 
However, when comparing BigDecimal values, assertEquals() can sometimes fail if the scales differ(ex. 50.5 vs 50.50).
Consider using different way instead, which ignores scale differences.

```suggestion
assertTrue(result.compareTo(new BigDecimal("50.5")) == 0)
```

**Review by Jeffrey-Hong1005 — COMMENTED** [review]
Review for AppTest code
1. Repeatation management
2. Big Decimal value

File: src/test/java/itm/oss/splitter/AppTest.java (hunk 3 above)
Comment (Kimhyewon0621) [reply src/test/java/itm/oss/splitter/AppTest.java]: Thanks for pointing that out redundant problem.
That was a clear issue we must change it.
I've committed the suggested change.
Please approve if it looks good!

File: src/test/java/itm/oss/splitter/AppTest.java (hunk 4 above)
Comment (Kimhyewon0621) [reply src/test/java/itm/oss/splitter/AppTest.java]: Thanks for your suggestion.
The name 'testApp_Placeholder()' wasn't clear it's our mistake..!
I've renamed the method to testAppInitialization as you suggested.
Please review and approve if it looks to good.

Comment (yeonlimee2) [reply src/test/java/itm/oss/splitter/AppTest.java]: Naming is much clearer than before!
Thank you for your checks.

File: src/test/java/itm/oss/splitter/AppTest.java (hunk 3 above)
Comment (yeonlimee2) [reply src/test/java/itm/oss/splitter/AppTest.java]: I've confirmed that the redundancyn is deleted.
LGTM^^

File: src/test/java/itm/oss/splitter/AppTest.java (hunk 5 above)
Comment (mingyeonggg) [reply src/test/java/itm/oss/splitter/AppTest.java]: Thanks for the great suggestion. I agree with you about the stream setup duplication. I've extracted the duplication stream logic into `@BeforeEach` and `@AfterEach` as you suggested. 
Additionally, I also removed the `try-finally` blocks from all test methods since it is no longer necessary. This greatly improves test file's clarity and maintainability.
Please take a look again my pushed changes and let me know if meets your approval!

File: src/test/java/itm/oss/splitter/AppTest.java (hunk 6 above)
Comment (mingyeonggg) [reply src/test/java/itm/oss/splitter/AppTest.java]: Thanks for leaving critical suggestion. We missed the potential issue of scale differences between two. I've applied the suggestion to replace `assertEquals(new BigDecimal("50.5"), result);` with `assertTrue(result.compareTo(new BigDecimal("50.5")) == 0)` as you suggested. I pushed the change. Could you please check my commit again for approval?

File: src/test/java/itm/oss/splitter/AppTest.java (hunk 5 above)
Comment (Jeffrey-Hong1005) [reply src/test/java/itm/oss/splitter/AppTest.java]: LGTM!! Thanks for application

File: src/test/java/itm/oss/splitter/AppTest.java (hunk 6 above)
Comment (Jeffrey-Hong1005) [reply src/test/java/itm/oss/splitter/AppTest.java]: LGTM. Now it looks more comfortable to me. Thanks!

**Review by Jeffrey-Hong1005 — APPROVED** [review]

**Review by yeonlimee2 — APPROVED** [review]
All feedback has been applied properly.
Tests look good and the naming is clear now.

**Review by ERE252 — APPROVED** [review]

File: src/main/java/itm/oss/splitter/App.java (hunk 1 above)
Comment (AyeongKwon) [reply src/main/java/itm/oss/splitter/App.java:88]: Thank you for your kind suggestion! I fixed it.

File: src/main/java/itm/oss/splitter/App.java (hunk 2 above)
Comment (AyeongKwon) [reply src/main/java/itm/oss/splitter/App.java]: I agree with your point! Thanks for your kindness. I added and fixed the code!

File: src/main/java/itm/oss/splitter/App.java (hunk 1 above)
Comment (ERE252) [reply src/main/java/itm/oss/splitter/App.java:88]: Great, thanks for the quick update!

File: src/main/java/itm/oss/splitter/App.java (hunk 2 above)
Comment (ERE252) [reply src/main/java/itm/oss/splitter/App.java]: LGTM. Great job addressing the feedback.
